pandas/_libs/tslibs/tzconversion.pyx

```diff
@@ -483,6 +489,72 @@ cdef _get_utc_bounds(
     return result_a, result_b
 
 
+cdef _get_utc_bounds_zoneinfo(ndarray vals, tz, NPY_DATETIMEUNIT creso):
+    """
+    For each point in 'vals', find the UTC time that it corresponds to if
+    with fold=0 and fold=1. In non-ambiguous cases, these will match.
+
+    Parameters
+    ----------
+    vals : ndarray[int64_t]
+    tz : ZoneInfo
+    creso : NPY_DATETIMEUNIT
+
+    Returns
+    -------
+    ndarray[int64_t]
+    ndarray[int64_t]
+    """
+    cdef:
+        Py_ssize_t i, n = vals.size
+        npy_datetimestruct dts
+        datetime dt, rt, left, right, aware, as_utc
+        int64_t val, pps = periods_per_second(creso)
+        ndarray result_a, result_b
+
+    result_a = cnp.PyArray_EMPTY(vals.ndim, vals.shape, cnp.NPY_INT64, 0)
+    result_b = cnp.PyArray_EMPTY(vals.ndim, vals.shape, cnp.NPY_INT64, 0)
+
+    for i in range(n):
+        val = vals[i]
+        if val == NPY_NAT:
+            result_a[i] = NPY_NAT
+            result_b[i] = NPY_NAT
+            continue
+
+        pandas_datetime_to_datetimestruct(val, creso, &dts)
+        # casting to pydatetime drops nanoseconds etc, which we will
+        #  need to re-add later as 'extra''
+        extra = (dts.ps // 1000) * (pps // 1_000_000_000)
+
+        dt = datetime_new(dts.year, dts.month, dts.day, dts.hour,
+                          dts.min, dts.sec, dts.us, None)
+
+        aware = dt.replace(tzinfo=tz)
+        as_utc = aware.astimezone(utc_stdlib)
+        rt = as_utc.astimezone(tz)
+        if aware != rt:
```

Comment:
Might be sufficient to check the `dst` attribute when constructing with `fold=0/1`

```
In [1]: import datetime

In [2]: import zoneinfo

In [3]: before = datetime.datetime(2022, 11, 6, 1, fold=0, tzinfo=zoneinfo.ZoneInfo("US/Pacific"))

In [4]: after = datetime.datetime(2022, 11, 6, 1, fold=1, tzinfo=zoneinfo.ZoneInfo("US/Pacific"))

In [5]: before
Out[5]: datetime.datetime(2022, 11, 6, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='US/Pacific'))

In [6]: after
Out[6]: datetime.datetime(2022, 11, 6, 1, 0, fold=1, tzinfo=zoneinfo.ZoneInfo(key='US/Pacific'))

In [7]: before.dst()
Out[7]: datetime.timedelta(seconds=3600)

In [8]: after.dst()
Out[8]: datetime.timedelta(0)
```

Reply:
IIUC that would check for if the datetime is ambiguous, but i dont think that checks for non-existent

Reply:
inferring is only supported for the `ambiguous` parameter though?

I think from the ` for i in range(n):` loop below, looks like nonexistent times should just fall though to the `else` clause i.e. nonexistent times shouldn't be populated in `result_a` and `result_b`

Reply:
> looks like nonexistent times should just fall though to the else

For that to happen, `result_a[i]` or `result_b[i]` need to be NPY_NAT; this is where that is getting set.

Reply:
i dont think there is anything actionable here?